crates/swc_ecma_minifier/src/compress/optimize/unused.rs

```diff
@@ -382,10 +396,16 @@ impl Optimizer<'_> {
 
                             self.take_pat_if_unused(&mut p.value, None, is_var_decl);
                         }
-                        ObjectPatProp::Assign(AssignPatProp {
-                            key, value: None, ..
-                        }) => {
-                            self.take_ident_of_pat_if_unused(key, None);
+                        ObjectPatProp::Assign(AssignPatProp { key, value, .. }) => {
+                            if has_pure_ann {
```

Comment:
`has_pure_ann` only represents that `init` is pure. Consider that `init` isn't pure, but the `value` is pure

Reply:
I don't expect such case to exist in real world code

Reply:
The `value` is an independent and unpredictable expression which inits the value of `undefined` key. I think it has nothing to do with `init`

Reply:
It's not simply dropping `value`; it's calling `ignore_return_value` on the value because we can drop the initializer.

Reply:
Yeap.
But if the `init` can't be dropped while the `value` can, this logic will skip dropping `value` because `has_pure_ann` is `false`.

Reply:
> I don't expect such case to exist in real world code

Reply:
for example:
```js
function getDefaultTextInfo() {
  return { a: 'a', b: 'b' };
}

const { textInfo = getDefaultTextInfo() } = useAtomValue(DataAtom);
```

Reply:
It's very common but not generated by tool. The example was just from my private project:
```js
const { state = {} } = getRuntimeInfo();
```

`state` is unused after tree shaking, but it cannot be removed.

Reply:
This is not removable unless pure_getters is true. A member access can have a side effect.

Reply:
If you are talking about `pure_getters: true`, feel free to send a PR

Reply:
> This is not removable unless pure_getters is true. A member access can have a side effect.

It can't be removed while `pure_getters` is true.

See [playground](https://play.swc.rs/?version=1.7.14&code=H4sIAAAAAAAAA8vMLcgvKlGoVkhPLQkqzSvJzE31zEvLV6hVSCvKz1VQL4KIqVtzcSXn5xWDVCbqKCQp2CpU1wIV2aLp09C0BqvLz0nVy8lP10jUtAYAWiuesmIAAAA%3D&config=H4sIAAAAAAAAA31Vy47bMAy85ysCn%2Few2EMPvRUFeus3CIpEOUol0RCpbIJF%2Fr2047w2dC6BoyE5fGn0tVqvux257uf6Sz7lz1CBoO7hV0q%2FMWcoTAIGmwjeZgtbxeLqISd0LGwPctKBy5ZcjQN3bxd0R4fHCK9ZJpPTzMW29sBTZPp4%2F3ifo3YJkeBbXjmWGI73eTnMIxHdnY126Fsanbm2a0ZyLlRNqXfGKn7Ss88GMYEtLxBjycTC0EPVAjtMyQ4EZm%2BrEmWswNZIqFGMYGPwZqg4qHjxkSMW4XxGPVhvHHqlEz5WcBz3oLkJl7gVwgRKPRPsYdP6ftqRb96wt6lZVjjhMI1KslWibjESm9CK1sIzuNCDMzg397tnDKYCt1qe%2FXYYy8JM%2FgFIB5IlKjaDFneyCLJPS97hpWcsQVaZjwoue69VWaCXppoYg9LZsTNQOWrTrOCbg7GzTktnhhfaR9GDgRBkV5TQ9BnZbTVSPg6AQQFkvjZoW3UGzPUWLuDjhXgB%2F5EqWV%2Bw2SJb3i6jdMwbTC8IMvAW%2FQsDGQXjMlxFJQ7DMt6KB1kN8KpJowl4FgG5AIwmTTr6tBtyPSSi6RNuVJkYWhUUmOF6F2bwdFXzbEt%2FU4NHAa%2BigQFrvtflCtbxoyjXVjjmMUZnG6MMIrruQrS6%2FE5RbwI%2BP1njPp2fiB%2FdzejyGlzq6SL9vVf%2B1ek%2FN4d5Zv0GAAA%3D)

Reply:
I know

Reply:
The closed #9460 was from this.